CIP-optimized-uplc-serialization/README.md

```diff
@@ -0,0 +1,327 @@
+
+```hs
+data Data =
+      Constr Integer [Data]
+    | Map [(Data, Data)]
+    | List [Data]
+    | I Integer
+    | B ByteString
+```
+
+a more efficient implementation could be obtained by introducing some tags for the various construcors, each taking 3 bits:
+
+data constructor | binary | decimal
+-----------------|--------|--------
+Constr           | 000    | 0
+Map              | 001    | 1
+List             | 010    | 2
+I                | 011    | 3
+B                | 100    | 4
+
+directly followed by the respective serialized value; in particoular
+
+- ```Constr``` is followed by an **unsigned** integer and the a list of ```Data```
+- ```Map``` is followed by a list of ```Data``` pairs; where the paris are serialized as in the constant pair specification (only the value)
+- ```List``` is folloed by a list of ```Data``` elements
+- ```I``` is followed by a **signed** integer
+- ```B``` is followed by a ```ByteString``` 
+
+An important **alternative design** that should be considered and discussed with the comunity would see the ```Data``` type requiring only 4 tags which would then be:
+
+data tag  | binary | decimal
+----------|--------|--------
+data-pair | 00     | 0
+data-list | 01     | 1
+data-int  | 10     | 2
+data-bs   | 11     | 3
+
+and the constructors would map to:
+
+- ```Constr``` becomes a ```data-pair``` expecting an unsigned integer and a list of ```Data```
+- ```Map``` becomes a ```data-list``` containing 2 ```Data``` per element
+- ```List``` becomes a ```data-list``` containing 1 ```Data``` per element
+- ```I``` becomes a ```data-int``` tag expecting a signed integer
+- ```B``` becomes a ```data-bs``` tag expecting a ```ByteString```
+
+this design is unambigous since ```data-list``` that maps back to two of the constructors contains different values depending on what Constructor is intended.
+
+the only edge case for the ```data-list``` constructor is the empty list, but the problem can be easly solved by expecting one single bit only in the case of the empty list indicating the intended constructor.
+
+so the empty ```Map``` would serialize as
+```
+01 # data-list tag
+0  # list nil construcor ( empty list )
+0  # bit indicating Map data constructor
+```
+and the empty ```List``` as
+```
+01 # data-list tag
+0  # list nil construcor ( empty list )
+1  # bit indicating List data constructor
+```
+
+the second design also implies no more ```Data``` constructors will be added in the future, or that for those added it will be possible to find an unambigous representation using the ones specified.
+
+## Rationale
+
+This proposal suggest ways to reduce serialized scripts size.
+
+the changes where designed to keep the bit oriented style of flat minimizing the number of required bits for values where possible.
```

Comment:
The rationale section is a bit _thin_, though there's a bit of rationale sprinkled in various places of the specification. It'd be nice to have a summary of what motivates each design choices. This can most likely be done by reworking a bit how the specification is written (that is, write the specification without justifications, and move the justifications to the rationale section). 

For example, most of the "data serialization" section is about explaining what is currently wrong with the `Data` CBOR serialization. Instead, leave the specification focused to what the proposal is about, and move the rationale to this section. 
  
It'd be nice to consider some benchmarks? Taking some reference UPLC and serializing them side-by-side with the two methods to see how much gain one can expect.

Reply:
Thank you @KtorZ; will refactor as suggested ( and correct the various typos 😄 )